File: src/main/java/at/searles/parsing/combinators/RecognizerToReducer.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.PartialConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;

public class RecognizerToReducer<T> implements Reducer<T, T> {
    private final Recognizer recognizer;

    public RecognizerToReducer(Recognizer recognizer) {
        this.recognizer = recognizer;
    }

    @Override
    public boolean recognize(Environment env, ParserStream stream) {
        long preStart = stream.start();

        boolean status = recognizer.recognize(env, stream);

        if(status) {
            stream.setStart(preStart);
        }

        return status;
    }


    @Override
    public T parse(Environment env, @NotNull T left, ParserStream stream) {
        long preStart = stream.start();

        if(!recognizer.recognize(env, stream)) {
            return null;
        }

        stream.setStart(preStart);
        return left;
    }

    @Override
    public PartialConcreteSyntaxTree<T> print(Environment env, @NotNull T t) {
        return new PartialConcreteSyntaxTree<>(t, recognizer.print(env));
    }

    @Override
    public String toString() {
        return recognizer.toString() + "{x->x}";
    }
}

File: src/main/java/at/searles/parsing/combinators/RecognizerToReducer.kt
package at.searles.parsing.combinators

import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizer
import at.searles.parsing.Reducer
import at.searles.parsing.printing.PartialTree

class RecognizerToReducer<T>(private val recognizer: Recognizer) : Reducer<T, T> {
    override fun reduce(left: T, stream: ParserStream): T? {
        if(!stream.recognize(recognizer, false)) {
            return null
        }

        return left
    }

    override fun recognize(stream: ParserStream): Boolean {
        return stream.recognize(recognizer, false)
    }

    override fun print(item: T): PartialTree<T> {
        return PartialTree(item, recognizer.print())
    }

    override fun toString(): String {
        return recognizer.toString()
    }
}
